### PR TITLE
Updates to the Pretty Printing Infrastructure

### PR DESCRIPTION
**Pull request by ejgallego**
This is the implementation of https://github.com/coq/ceps/blob/master/text/009-unified-pretty-printing.md plus some additional cleanups and bugfixes.

This commit doesn't cover the advanced features proposed on the CEP as some more discussion and design is needed, in particular with the maintainers of externalization and the  notation system.

This PR also implements the client-side automatic printing width in CoqIDE feature.

The most important commit is "[pp] Make feedback the only logging mechanism."  , this unifies all Coq messaging to use a single datatype `std_ppcmds`. This seems to have several advantages as outlined in the commit log.

This PR supersedes PR #185 , see also the discussion there.

### CODE REVIEW

**Comment by psteckler** [comment]
This means that the <richpp> tags in the XML protocol, introduced in 8.6, will disappear? Or merely deprecated? If it goes away, we'll see an XML version of the new format you propose in the CEP?

Conceivably, for PG, the new format could be processed by an XSLT transformer inside Emacs, though that's a non-trivial amount of work, I'd think.

**Comment by ejgallego** [comment]
Hi Paul,

this PR will maintain protocol-level compatibility, that is to say, it will still send the `<richpp>` by default. XML clients may request to get the `stdpp_cmds` if the want (note that this format precedes `richpp`) but that is, no more evolution on that side.

The CEP is currently out of date, I'll try to update it soon. After much thinking, I am convinced now that what I want to do cannot be done with the current Coq printing infrastructure, so I am working on a new printer that avoids externalization and has a different set of features. That printing stack will be only available to SerAPI clients, but I make write a small plugin exporting its functionality throu a vernacular if there is demand.

**Comment by ejgallego** [comment]
Regarding emacs, if you want to process `std_cmdpp`, it is not really different from processing `richpp`, it is a very simple format, you just need a bit of care with indentation.

For the new printer, maybe TeX is the best output option for emacs, but HTML/XML could also work.

**Comment by psteckler** [comment]
The new printing stack is different from your discussion at the WG (which I just got around to watching)?

**Comment by ejgallego** [comment]
It implements the same features, but in the WG I was thinking of a different design, discussion and feedback there made me fully change the implementation approach.

**Comment by silene** [comment]
There are a few functions in `ide_slave.ml` that now return empty strings. I don't know where these strings are used, but if they are not, it would be better to just change the signature of these functions.

**Comment by ejgallego** [comment]
Hi @silene, good point, thanks!

The strings were used to capture the stdout/stderr that happened during the execution command, and as far as I can tell they are never used now as all the messages for users are routed throu feedback.

If stderr produces output from Coq now, it will go to the usual descriptor for CoqIDE, I should document this.

Unfortunately, it is not possible to fully remove this field from the protocol without breaking compatibility, so XML will still have to carry the empty tag.

**Comment by ejgallego** [comment]
@silene unfortunately removing the empty return parameter would break the protocol, and workarounding that adds IMO a lot of complexity to the code. So unless we provide a protocol versioning mechanism I'm gonna let things as they are I think.

**Comment by silene** [comment]
Which protocol are you talking about? Your patch is already touching about 25 .mli files. I don't see how fixing the signatures of these functions so that they match their new semantics is going to make things any worse.

**Comment by ejgallego** [comment]
I am talking about the XML protocol.

I we change the signature of these functions (defined in `interface.mli`), either the XML protocol will change in a non backwards compatible way, breaking all clients, or it would require a messy workaround in `xmlprotocol.ml`. IMO returning `""` is the easiest way to maintain compatibility.

**Comment by ejgallego** [comment]
The signature of these functions is used later to do "type-based" XML serialization.

**Comment by silene** [comment]
> it would require a messy workaround in `xmlprotocol.ml`

Given that the `xmlprotocol.ml` file now explicitly supports two different protocols (the one for 8.6 and the one for 8.7), I don't think it matters much. And I want to stress it again, keeping the old signatures when the semantics have changed does not qualify as preserving backward compatibility.

**Comment by ejgallego** [comment]
Personally I am reluctant to change the protocol just for that, specially when in 8.6, the field was always filled with `""`.

IMHO, should we increase the protocol version number (which is a big pain for clients), then I would go ahead and try to improve some other aspects of the current protocol.

**Comment by ejgallego** [comment]
So that would go IMO in a different PR, "modifications to the XML protocol".

**Comment by ejgallego** [comment]
Well, somebody broke trunk and the the whole test queue by committing untested changes, closing and reopening so the PR gets back to life.

**Comment by ejgallego** [comment]
Code-wise this PR is ready, reviews and feedback is welcome.

A couple of contribs need to be fixed, but that will happen soon.

**Comment by ejgallego** [comment]
Pending on math-comp/math-comp#108

**Comment by ejgallego** [comment]
Hi folks, this PR is risking going stale so I would like to propose it for merging in the next few days or either take a different action path.

Having a look at https://coq.inria.fr/cocorico/CoqDevelopment/Maintainers , I guess that @silene @herbelin  @ppedrot @mattam82 @maximedenes and @gares are the relevant people here.

Do you have anything against pushing this to trunk? Do you buy my arguments for considering this a positive change?

Thanks,
E.

**Comment by maximedenes** [comment]
@silene are you ok with keeping the empty strings and to integrate such refactorings in bigger changes in the protocol to break less often (I guess it is @ejgallego 's intention).

**Comment by maximedenes** [comment]
I'd like to merge this. @silene @herbelin @ppedrot @mattam82 or @gares, any objections?

**Comment by silene** [comment]
It depends whether you can promise me that Coq 8.7 will not be shipped with such a broken interface.

**Comment by gares** [comment]
My impression is that the "string" we are talking about was there to send to the UI output the prover was writing to stdout/err.  Since 8.5 Coq uses the feedback mechanism, but 1) not the entire code base was ported to that; 2) a plugin can always write `Printf.eprintf ...`.  Such output is a second class citizen, since it is not attached (as feedback is) to a particular state_id (line in the input text).  So CoqIDE simply puts it in the output window trying to do a best effort service.

IMO this PR should not get rid of this safety measure (the redirection of standard file descriptors) that fills in the string (that is now empty).  @ejgallego seems to suggest that there is another countermeasure in a comment but I've not time to find where it is.  Is it the case?

Let me recall that Proof General still uses a "communication protocol" based on stdin/stdout/stderr so such messages would be displayed in PG but not CoqIDE.

Once we will have a documented API, we can pretend plugin writers don't print to stderr/stdout and expect their prints to work.

**Comment by ppedrot** [comment]
I think I already complained about this somewhere, but I can't find where, so let me reiterate it here. I think making the pp_stdcmds type completely transparent is a **Very Bad Idea** ©. It prevents using things like smart constructors and makes maintenance harder. @ejgallego, could you instead provide a `Pp.repr` function that gives a view to the actual inductive datatype? It does not matter if this is mere identity, I really think that representation should be encapsulated and ensured to appear in the higher layers.

**Comment by ejgallego** [comment]
I answer point by point:

 - @silene IMO the interface is not broken, or at least is as broken as it in 8.6, as in 8.6 it is always the case that here we have the empty string. Again I insist this is a matter of **cost/benefit**. The cost of changing the type is quite high at it will mean that suddenly all the clients using the XML protocol will be broken. I don't have a problem breaking clients, however, in this case I see 0 benefit other than cleaning up 3 lines of code in a private interface. So IMHO the tradeoff is clear here, but if others thing we should push a new version of the protocol for such a minor benefit I will proceed to do so.
- @gares The patch in its current state will indeed make code that uses `{e,}printf` to output to the real descriptors, so you see such output in the console, etc... No plugin is actually using the interface. I am not sure messing with `std*` was the right thing to do, in particular as it only happened between calls which may be broken in the proccess case? Thus my policy was to remove such mechanism as require that all plugins use the new API (which I see documented, what is missing here) Debug operations, etc... (think of `-debug`) can write to stderr, and clients can handle their descriptors in a standard way. I think emacs does, but @psteckler should confirm. Again this can be reverted back if desired, but I'd advise against this hybrid model.
- @ppedrot I still fail to see why it is such a bad idea, in particular when this type is a clear candidate for display plugins to manipulate. I think it is pretty clear in the API that you are supposed to use the functions, otherwise you may pay the price. Moreover, a more fundamental problem about the `repr` idea is that elements of the `pp_stdcmd` type have to be faithfully sent from/to workers. Failure to do so in the past led to quite a bit of mess, including the main one of introducing a duplicate datatype that then `master` threads could not manipulate, see for example (the totally legitimate complain https://github.com/coq/coq/blob/trunk/stm/asyncTaskQueue.ml#L301 The datatype could be marked as private but I see that as a mere annoyance, better to be sure we have true 1:1 marshalling .

So in short, my preliminary assessment of the comments are:
- @silene 's : Soft reject: bad cost/benefit ratio.
- @gares : Soft reject based on design / architectural / code simplicity issues. (Mainly what happens when output to std* happens between commands being unspecified, interaction with -debug)
- @ppedrot : Reject, making the datatype public is an essential design choice of this PR, reverting on that policy may bring back some of the old problems. Also, I perceive a high cost/benefit ratio.

Looking forward to changing my mind!

**Comment by ejgallego** [comment]
I've updated my reply to reflect that @silene complaints about the interface are legitimate, indeed we should squash his change in the next update of the XML protocol. Is there a roadmap for it so I can add this item and doesn't get forgotten?

**Comment by ppedrot** [comment]
I really don't get why you're not willing to make the datatype opaque. If you provide a pair of conversion functions from and to the transparent representation, it is equally expressive as the system you propose, with the additional safety belt that you can precisely encapsulate the parts which rely on the internal representation. Having everything transparent is bad practice and, even in the documentation you wrote, you strongly discourage people to use the non-abstract representation. Why not enforce this through the compiler? Am I the only people caring for static guarantees here? Can't we summon some OCaml expert?

**Comment by ejgallego** [comment]
@ppedrot it is IMHO a matter of design not about types. There is an essential fact, even in the current design, the internal representation **leaks** here when you involve the client.

In the current state of affairs, it leaks throu `Richpp.repr` and clients have to deal with that representation (which is a bit odd, but for other reasons). So I think it is better to expose what the client will find from the start.

It could also be argued that having everything opaque is equally bad practice. Each case has a different set of constraints I guess. Would we had a standard serialization method in Coq, so we had `Pp.t` and `Pp.sexp_of_t` etc... the case for a private data type would be more convincing, but even so, I can perfectly envision plugins willing to manipulate documents, so I favor public right now.

I am not sure what an OCaml expert could contribute here as we are really talking about internal design choices of the Coq printing path as outlined in CEP009, but feel free to invite some. Maybe I am missing something but I don't really see the difference, specially when Proof General and other clients need to deal with the internal representation anyways.

> even in the documentation you wrote, you strongly discourage people to use the non-abstract representation

Eh? Quoting the commit:
```ocaml
(* The datatype is public to allow serialization or advanced uses,      *)
(* regular users are _strongly_ encouraged to use the top-level         *)
(* functions to manipulate the type.                                    *)
```
maybe you mean the wording is not clear enough? I don't think it was my intention to discourage people to use the non-abstract representation but quite the opposite.

> Am I the only people caring for static guarantees here?

What is the static guarantee that an opaque representation would provide here?

**Comment by gares** [comment]
I've the impression we are all neatpicking.  The benefits of this PR is much higher than the drawbacks it introduces.

**Comment by ppedrot** [comment]
A purely quantitative argument that may convince you better: if 99% of the code is only caring about building up formatting for printing, why contaminate the representation of such a pervasive type as `std_ppcmds` while you only need this for like 1% in the marshalling API? People writing plugins will start using the transparent representation, and then, we're dead.

Remember, developpers don't give a shit about good practices documentation (and I include myself in this statement, see e.g. how I like to abuse OCaml internal representations).

**Comment by ejgallego** [comment]
> A purely quantitative argument that may convince you better: if 99% of the code is only caring about building up formatting for printing, why contaminate the representation of such a pervasive type as std_ppcmds while you only need this for like 1% in the marshalling API?

Well quantitative arguments are always tricky in the sense that this 1% API is called basically at every single message the system produces. The famous 1% of the code is responsible for 99% of the (bugs? :D)

Another point I don't understand is what is your definition of "contamination". What is exactly being contaminated here?
>People writing plugins will start using the transparent representation, and then, we're dead.

Why should they? I agree that maybe the documentation should be more explicit, I'll fix that. Note however that there are some legitimate use cases for doing so! (For example a plugin that converts Coq documents to TeX)

> Remember, developers don't give a shit about good practices documentation (and I include myself in this statement, see e.g. how I like to abuse OCaml internal representations).

I am a bit confused here, how can this choice be abused, it is not like we are exposing something unsafe here?

**Comment by ppedrot** [comment]
To be more constructive, I can provide a patch atop of this branch to show that the alleged burden is ridiculous.

> What is the static guarantee that an opaque representation would provide here?

That people have to go through the abstract API. If ever we want to change or extend the datatype, and opaque API ensures that only people writing marshalling functions are affected. If the type is transparent, people will start abusing this "feature" and we'll have to change like hundreds to thousands of lines of code.

**Comment by ejgallego** [comment]
> To be more constructive, I can provide a patch atop of this branch to show that the alleged burden is ridiculous.

I know what the burden is and I guess we all know what such a patch would look like. I think it may be more effective time-wise to reach an agreement before.

> That people have to go through the abstract API. If ever we want to change or extend the datatype, and opaque API ensures that only people writing marshalling functions are affected. If the type is transparent, people will start abusing this "feature" and we'll have to change like hundreds to thousands of lines of code.

Ok so here we are with the core point. So, IIUC your proposal:
- how can people not to abuse the representation? They can always project to the representation and have the fun anyways.
- how does it protect us wrt to evolutions of the infrastructure. For instance, in the future is planned to add a new constructor to the datatype representing notations. In both cases, we would break all the clients using the repr.

I agree that the point of opaque datatypes is to protect the internal representation, I agree. But IMO, now the internal and external representation is the same. I cannot see a case yet we would need to represent documents using octa-trees or some other fancy structure, but the time may of course come.

In that model, the current datatype would indeed become the view and we would have to make the datatype private. So yeah, I am more convinced about your proposal now.

**Comment by ejgallego** [comment]
In fact in some sense what you propose was left for future work but it may be worth addressing it. I have been thinking for instance about having a kind of normal forms for documents.

Indeed, using your approach it could make sense, however I was thinking about maybe a plugin taking care of this?

I need to think a bit more, but I will update my answer to reflect that change of opinion.

**Comment by psteckler** [comment]
It looks like old PG is able to ignore output on `stderr`, though I can see it in the `*coq*` buffer.

In PG/xml, `stderr` is redirected to `/dev/null`, because pre-v25 Emacs doesn't provide a clean way to separate `stdin` and `stderr`.

**Comment by psteckler** [comment]
Re versioning of the XML protocol: there is a versioning mechanism available via the `About` call. The version is a date YYYYMMDD. So far, there have two versions, for 8.5 and 8.6. PG/xml is already using the version information. So as long as changes are guarded by the version number, and the changes are made clear, clients can handle them.

**Comment by ejgallego** [comment]
Ok @ppedrot , after some thinking I've became convinced, it doesn't cost a lot to play on the safe side here for now. Wanna have a look at ce0a1df ? Any other change you would do there?

**Comment by ejgallego** [comment]
Ok so the current status is:
- I've added a note about the leftover types, it should be pretty clear now. I propose we include @silene's request in a next revision of the XML protocol. I am sure we can collect quite a few wishlists about the protocol and do the version change at once. (Maybe github's wiki would be a good place to build that wishlist?)
- I've done as I understood @ppedrot suggested, if he agrees with the change that should be closed.

A general policy this PR is imposing (documented in the commit log) is that it is **forbidden* for the core of Coq to access the console. The console is usually owned by the upper layer and as such, this layer has complete control. Of course, we all can still use the occasional hack for debugging, but as a general principle I believe this should be the general choice.

I did not put any mechanism in place to enforce this policy, I think that'd be overkill, but indeed I revised the whole codebase, starting with #179 to ensure that no access to the console remains, except when the -debug flag is given.

So if we all agree on that we should be ready to go.

**Comment by ppedrot** [comment]
LGTM.

**Comment by maximedenes** [comment]
Merged in trunk to avoid huge maintenance cost (and because no fundamental opposition). Feel free to open a PR dedicated to the XML protocol issue.

Thanks for this work, Emilio!

**Comment by ejgallego** [comment]
Great thanks, @silene remark is queued in the TODO list for before 8.7, it seems indeed we may update some other parts of the xml protocol so I'll try to be sure that cleanup is squashed there. WRT Enrico comments IMVHO I think it is sensible for now to maintain the outlined console policy, however other alternatives should be possible so we should discuss this (maybe at a WG) if people is interested.

**Comment by ppedrot** [comment]
@ejgallego: It seems your refactoring broke something in the Query call to the STM. See [this bug](https://coq.inria.fr/bugs/show_bug.cgi?id=5417) for more details.
